walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,81 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> requiredEvents = findAttendeeEvents(events, request.getAttendees());
+    List<TimeRange> optionalEvents = findAttendeeEvents(events, request.getOptionalAttendees());
+
+    List<TimeRange> allAttendees = findOpenTimes(requiredEvents, optionalEvents, request);
+    List<TimeRange> onlyOptional = findOpenTimes(Arrays.asList(), optionalEvents, request);
+    List<TimeRange> onlyMandatory = findOpenTimes(requiredEvents, Arrays.asList(), request);
+
+    if (allAttendees.size() > 0) {
+        return allAttendees;
+    } else if (requiredEvents.size() > 0) {
+        return onlyMandatory;
+    } else {
+        return onlyOptional;
+    }
+  }
+
+  private List<TimeRange> findOpenTimes(List<TimeRange> mandatoryAttendeeMeetings, 
+                            List<TimeRange> optionalAttendeeMeetings, MeetingRequest request) {
+
+    ArrayList<TimeRange> options = new ArrayList<TimeRange>();
+    int earliestAvailable = TimeRange.START_OF_DAY;
+
+    List<TimeRange> allMeetings = new ArrayList<TimeRange>(mandatoryAttendeeMeetings);
+    if (optionalAttendeeMeetings != null) {
```

Comment:
Question: Would optionalAttendeeMeetings to be null? Seems like we always return an empty List by default.
If that is the case, you can remove the if statement.